File: algorithms/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        if not s:
            return True
        a = {
            '(': ')',
            '[': ']',
            '{': '}',
        }
        stack = list()
        for i in s:
            if i in a:
                stack.append(i)
                continue
            if not stack:
                return False
            if a[stack.pop()] != i:
                return False
        return not bool(stack)

File: algorithms/0020-valid-parentheses.go
package algorithms

func isValid(s string) bool {
	if s == "" {
		return true
	}
	a := map[byte]byte{
		'(': ')',
		'[': ']',
		'{': '}',
	}
	stack := make([]byte, 0)
	for _, i := range s {
		j := byte(i)
		if j == '(' || j == '[' || j == '{' {
			stack = append(stack, j)
		} else if len(stack) > 0 && a[stack[len(stack)-1]] == j {
			stack = stack[:len(stack)-1]
		} else {
			return false
		}
	}
	return len(stack) == 0
}
